packages/next/next-server/server/image-optimizer.ts

```diff
@@ -281,30 +284,46 @@ export async function imageOptimizer(
     const optimizedBuffer = await transformer.toBuffer()
     await promises.mkdir(hashDir, { recursive: true })
     const extension = getExtension(contentType)
-    const filename = join(hashDir, `${expireAt}.${extension}`)
+    const etag = getHash([optimizedBuffer])
+    const filename = join(hashDir, `${expireAt}.${etag}.${extension}`)
     await promises.writeFile(filename, optimizedBuffer)
-    res.setHeader('Content-Type', contentType)
-    res.end(optimizedBuffer)
+    sendResponse(req, res, contentType, optimizedBuffer)
   } catch (error) {
-    server.logError(error)
-    if (upstreamType) {
-      res.setHeader('Content-Type', upstreamType)
-    }
-    res.end(upstreamBuffer)
+    sendResponse(req, res, upstreamType, upstreamBuffer)
   }
 
   return { finished: true }
 }
 
+function sendResponse(
+  req: IncomingMessage,
+  res: ServerResponse,
+  contentType: string | null,
+  buffer: Buffer
+) {
+  const etag = getHash([buffer])
+  res.setHeader('Cache-Control', 'public, max-age=0, must-revalidate')
```

Comment:
@kodiakhq @Timer 

I don't understand this line - why is the cache max-age set to 0? Does this affect images?  

I thought images would be 60 seconds as per https://nextjs.org/docs/basic-features/image-optimization#caching

Reply:
I agree it's a bit confusing, it caught me off guard as well.

What I get from it is that there are two different forms of caching: on the process hosting NextJS (server) and browser (client) side.

Here, the client side is instructed to always check if there's a newer version of the image available. Which is necessary with the introduction of etags to allow the NextJS process to invalidate images at any time.

On the server side, NextJS has its own cache for the generated images. The cached image will be served according to the `max-age` header value when retrieving the image, defaulting to 60 seconds if the header was not supplied.